File: frontend_v2/models/memory.ts
import { BaseEntity, ID } from './common';

export type MemoryTag = 'important' | 'goal' | 'challenge' | 'achievement' | 'relationship' | 'health' | 'work' | 'personal' | 'custom';

export interface Memory extends BaseEntity {
  userId: ID;
  content: string;
  title: string;
  tags: MemoryTag[];
  isPinned: boolean;
  source?: 'chat' | 'reflection' | 'manual';
  relatedConversationId?: ID;
  emotion?: string;
}

export interface CreateMemoryRequest {
  title: string;
  content: string;
  tags?: MemoryTag[];
  relatedConversationId?: ID;
}

export interface UpdateMemoryRequest {
  title?: string;
  content?: string;
  tags?: MemoryTag[];
  isPinned?: boolean;
}

export interface MemoryFilter {
  tags?: MemoryTag[];
  startDate?: number;
  endDate?: number;
  searchQuery?: string;
}

export type MemoryNodeType =
  | 'subject'
  | 'experience'
  | 'emotion'
  | 'trigger'
  | 'thought'
  | 'behaviour'
  | 'topic'
  | 'memory';

export interface MemoryNode {
  id: ID;
  type: MemoryNodeType;
  label: string;
  title: string;
  preview: string;
  properties: Record<string, any>;
  editableFields: string[];
  enumFields: Record<string, string[]>;
  embeddingSynced?: boolean;
  updatedAt?: string;
}

export interface MemoryNodeListResponse {
  nodes: MemoryNode[];
  nodeType: MemoryNodeType;
  total: number;
}

export interface MemoryGraphRelation {
  id: ID;
  sourceId: ID;
  sourceType: string;
  sourceTitle: string;
  targetId: ID;
  targetType: string;
  targetTitle: string;
  relationType: string;
  label: string;
  confidence?: number;
}

export interface MemoryGraphRelationListResponse {
  relations: MemoryGraphRelation[];
  total: number;
}

export interface UpdateMemoryNodeRequest {
  properties: Record<string, any>;
}

export interface MemoryResetResponse {
  reset: boolean;
  nodesDeleted: number;
  sessionsDeleted: number;
  userRelationshipsDeleted: number;
  pgvectorRowsDeleted: number;
  userDeleted: number;
}

export interface MemoryPurgeResponse {
  purged: boolean;
  nodesDeleted: number;
  sessionsDeleted: number;
  userDeleted: number;
  pgvectorRowsDeleted: number;
}

File: agentic/gateway/model/memory.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


MemoryNodeType = Literal[
    "subject",
    "experience",
    "emotion",
    "trigger",
    "thought",
    "behaviour",
    "behavior",
    "topic",
    "memory",
    "person",    # deprecated alias for "subject" — normalised in gateway/service/memory_graph.py
]


class MemoryNode(BaseModel):
    model_config = ConfigDict(extra="allow")

    id: str
    type: str
    label: str
    title: str
    preview: str = ""
    properties: dict[str, Any] = Field(default_factory=dict)
    editable_fields: list[str] = Field(default_factory=list)
    enum_fields: dict[str, list[str]] = Field(default_factory=dict)
    embedding_synced: bool | None = None
    updated_at: str | None = None


class MemoryNodeListResponse(BaseModel):
    nodes: list[MemoryNode]
    node_type: str
    total: int


class MemoryGraphRelation(BaseModel):
    id: str
    source_id: str
    source_type: str
    source_title: str
    target_id: str
    target_type: str
    target_title: str
    relation_type: str
    label: str
    confidence: float | None = None


class MemoryGraphRelationResponse(BaseModel):
    relations: list[MemoryGraphRelation]
    total: int


class MemoryNodeUpdateRequest(BaseModel):
    user_id: str
    properties: dict[str, Any]


class MemoryNodeUpdateResponse(BaseModel):
    node: MemoryNode
    updated: bool
    pgvector_synced: bool | None = None


class MemoryNodeDeleteResponse(BaseModel):
    deleted: bool
    archived: bool
    pgvector_archived: int = 0


class MemoryResetResponse(BaseModel):
    reset: bool
    nodes_deleted: int = 0
    sessions_deleted: int = 0
    user_relationships_deleted: int = 0
    pgvector_rows_deleted: int = 0
    user_deleted: int = 0


class MemoryPurgeResponse(BaseModel):
    purged: bool
    nodes_deleted: int = 0
    sessions_deleted: int = 0
    user_deleted: int = 0
    pgvector_rows_deleted: int = 0


class MemorySessionPurgeRequest(BaseModel):
    message_ids: list[str] = Field(default_factory=list)

    @field_validator("message_ids", mode="before")
    @classmethod
    def default_message_ids(cls, value: Any) -> list[str] | Any:
        if value is None:
            return []
        return value


class MemorySessionPurgeResponse(BaseModel):
    purged: bool
    relationships_deleted: int = 0
    sessions_deleted: int = 0
    nodes_deleted: int = 0
    pgvector_rows_deleted: int = 0
    messages_processed: int = 0
